### DISCUSSION TITLE
Easier way to listen to DB tables via new plugin

### DISCUSSION THREAD

**Discussion by jan-tennert** [supabase-community/supabase-kt#379] [Ideas]
Continuation of https://github.com/supabase-community/supabase-kt/issues/369

**Comment by jan-tennert** [comment]
@shubhamsinghshubham777 So for the single-value flow this is certainly possible as we can just apply a complex query for the intial data and then just use the primary key as the realtime filter. For the multi-value flow this is a little bit trickier. I don't think it's a good idea to make n amount of channels and then combine the n flows. This leaves us two options:
- Don't provide any realtime filter and receive all data. Then we could apply the filters locally (which would require changes in the Postgrest internals and might be more expensive when receiving a lot of data, but would have all of the features)
- Do the same like for the single-value variant: Apply the filter in the initial request and then just use the initial data's primary keys as a filter (this obivously means that no inserts are supported and only a max of 100 entries can be checked)

**Reply by shubhamsinghshubham777** [reply]
> @shubhamsinghshubham777 So for the single-value flow this is certainly possible as we can just apply a complex query for the intial data and then just use the primary key as the realtime filter. For the multi-value flow this is a little bit trickier. I don't think it's a good idea to make n amount of channels and then combine the n flows. This leaves us two options:
> - Don't provide any realtime filter and receive all data. Then we could apply the filters locally (which would require changes in the Postgrest internals and might be more expensive when receiving a lot of data, but would have all of the features)
> - Do the same like for the single-value variant: Apply the filter in the initial request and then just use the initial data's primary keys as a filter (this obivously means that no inserts are supported and only a max of 100 entries can be checked)

@jan-tennert I think such complex filters would only make sense for the single data flows. Like you said, the complexity could grow multiple folds for listFlows.

Generally, more than lists, observation is done for a single data only. We can move forward with this for now 👍🏻

**Reply by shubhamsinghshubham777** [reply]
Just to reiterate, we're going to allow complex filters for single data flows and simple filters for list flows. Right?

**Reply by jan-tennert** [reply]
Yes, we can just reuse the PostgrestFilterBuilder for single data flows:
```kotlin
val dataFlow: Flow<Message> = table.dataFlow(
   filter = {
       Message::createdAt eq Clock.System.now()
       and {
            Message::creatorId eq "some-id"
            Message::content isIn listOf("some-content")
        }
    },
    primaryKey = {
        PrimaryKey("id", it.id.toString())
    }
)
dataFlow.collect {
    println(it)
}
```
